File: tr.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <portaudio.h>
#include <sndfile.h>
#include <pthread.h> // Threading için
#include <unistd.h>  // sleep için

//sudo apt-get update
//sudo apt-get install portaudio19-dev libsndfile1-dev
//gcc audio_app.c -o audio_app -lportaudio -lsndfile -lm -lpthread

// ==================
// ANSI Renk Kodları
// ==================
#define RESET   "\033[0m"
#define BOLD    "\033[1m"
#define UNDERLINE "\033[4m"
#define BLACK   "\033[30m"
#define RED     "\033[31m"
#define GREEN   "\033[32m"
#define YELLOW  "\033[33m"
#define BLUE    "\033[34m"
#define MAGENTA "\033[35m"
#define CYAN    "\033[36m"
#define WHITE   "\033[37m"
#define BRIGHT_BLACK "\033[90m"
#define BRIGHT_RED "\033[91m"
#define BRIGHT_GREEN "\033[92m"
#define BRIGHT_YELLOW "\033[93m"
#define BRIGHT_BLUE "\033[94m"
#define BRIGHT_MAGENTA "\033[95m"
#define BRIGHT_CYAN "\033[96m"
#define BRIGHT_WHITE "\033[97m"

#define GET_COLOR(color_code) color_code

// ==================
// Genel Ayarlar
// ==================
#define SAMPLE_RATE         44100
#define FRAMES_PER_BUFFER   512
#define NUM_CHANNELS        1
#define RECORDING_FILENAME  "kaydedilen_karisik_ses_c.wav"
#define PITCH_SHIFT_STEPS   -4

// DEĞİŞİKLİK: Sabit DURATION_SECONDS kaldırıldı.

// ==================
// Realtime Data Yapısı
// ==================
typedef struct {
    float *buffer;
    long buffer_size;
    pthread_mutex_t mutex;
    pthread_cond_t cond_data_available;
    pthread_cond_t cond_buffer_empty;
    long read_idx;
    long write_idx;
    bool terminate;
} RealtimeBuffer;

RealtimeBuffer inputBuffer;
RealtimeBuffer outputBuffer;

// ==================
// Fonksiyon Prototipleri
// ==================
void initialize_realtime_buffer(RealtimeBuffer *rb, long size);
void destroy_realtime_buffer(RealtimeBuffer *rb);
bool write_to_buffer(RealtimeBuffer *rb, const float *data, long frames);
bool read_from_buffer(RealtimeBuffer *rb, float *data, long frames);

int paCallback(const void *inputBufferPtr, void *outputBufferPtr,
               unsigned long framesPerBuffer,
               const PaStreamCallbackTimeInfo* timeInfo,
               PaStreamCallbackFlags statusFlags,
               void *userData);

void *realtime_processor_thread(void *arg);
void simple_pitch_shift(float *data, long num_frames, int sample_rate, int n_steps);
void record_process_play_save_mode();
void realtime_mode();
void display_menu();
void clear_input_buffer();

// ==================
// Main Fonksiyonu
// ==================
int main() {
    PaError err;
    int choice;

    err = Pa_Initialize();
    if (err != paNoError) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"PortAudio hatası: %s\n"RESET, Pa_GetErrorText(err));
        return 1;
    }

    while (true) {
        display_menu();
        printf(GET_COLOR(BRIGHT_WHITE)"Lütfen bir seçenek girin "GET_COLOR(CYAN)"(1-3)"GET_COLOR(BRIGHT_WHITE)": "RESET);

        if (scanf("%d", &choice) != 1) {
            printf(GET_COLOR(RED)"Geçersiz giriş. Lütfen bir sayı girin."RESET"\n");
            clear_input_buffer();
            sleep(1);
            continue;
        }
        clear_input_buffer();

        if (choice == 1) {
            printf(GET_COLOR(MAGENTA)BOLD">>> Seçim: Kaydet & İşle Modu <<<"RESET"\n\n");
            record_process_play_save_mode();
            printf(GET_COLOR(BRIGHT_YELLOW)"%s\n"RESET, "==================================================");
        } else if (choice == 2) {
            printf(GET_COLOR(MAGENTA)BOLD">>> Seçim: Gerçek Zamanlı Mod <<<"RESET"\n\n");
            realtime_mode();
            printf(GET_COLOR(BRIGHT_YELLOW)"%s\n"RESET, "==================================================");
        } else if (choice == 3) {
            printf(GET_COLOR(BRIGHT_RED)"Programdan çıkılıyor. Hoşça kalın!"RESET"\n");
            break;
        } else {
            printf(GET_COLOR(RED)"Geçersiz seçenek. Lütfen "GET_COLOR(CYAN)"1, 2"GET_COLOR(RED)" veya "GET_COLOR(RED)"3"GET_COLOR(RED)" girin."RESET"\n");
            sleep(1);
        }
    }

    err = Pa_Terminate();
    if (err != paNoError) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"PortAudio hatası: %s\n"RESET, Pa_GetErrorText(err));
    }

    return 0;
}


// ==================
// 1. KAYDET → İŞLE → DİNLE → KAYDET Modu Fonksiyonu
// ==================
void record_process_play_save_mode() {
    PaStream *stream = NULL;
    PaError err;
    int duration_seconds; // DEĞİŞİKLİK: Süre için yerel değişken

    // DEĞİŞİKLİK: Kullanıcıdan kayıt süresini al
    printf(GET_COLOR(BRIGHT_WHITE)"Lütfen kayıt süresini saniye cinsinden girin: "RESET);
    if (scanf("%d", &duration_seconds) != 1) {
        printf(GET_COLOR(RED)"Geçersiz giriş. Lütfen bir sayı girin."RESET"\n");
        clear_input_buffer();
        return;
    }
    clear_input_buffer();

    if (duration_seconds <= 0) {
        printf(GET_COLOR(RED)"Geçersiz süre. Süre 0'dan büyük olmalıdır."RESET"\n");
        return;
    }

    long num_frames = SAMPLE_RATE * duration_seconds * NUM_CHANNELS;
    float *recorded_samples = (float*) calloc(num_frames, sizeof(float));

    if (!recorded_samples) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"Bellek ayırma hatası!\n"RESET);
        return;
    }

    printf(GET_COLOR(BRIGHT_CYAN)"[KAYIT] Konuşmaya başlayın (%d saniye)..."RESET"\n", duration_seconds);

    err = Pa_OpenDefaultStream(&stream, NUM_CHANNELS, 0, paFloat32, SAMPLE_RATE, paFramesPerBufferUnspecified, NULL, NULL);
    if (err != paNoError) goto error_record;

    err = Pa_StartStream(stream);
    if (err != paNoError) goto error_record;

    err = Pa_ReadStream(stream, recorded_samples, num_frames);
    if (err != paNoError) goto error_record;

    err = Pa_StopStream(stream);
    if (err != paNoError) goto error_record;
    err = Pa_CloseStream(stream);
    if (err != paNoError) goto error_record;
    stream = NULL;

    printf(GET_COLOR(BRIGHT_CYAN)"[KAYIT] Tamamlandı.\n"RESET);
    printf(GET_COLOR(BRIGHT_YELLOW)"[İŞLEME] Ses işleniyor...\n"RESET);

    simple_pitch_shift(recorded_samples, num_frames, SAMPLE_RATE, PITCH_SHIFT_STEPS);

    for (long i = 0; i < num_frames; ++i) {
        float noise = ((float)rand() / (float)RAND_MAX) * 0.006f - 0.003f;
        recorded_samples[i] += noise;
        if (recorded_samples[i] > 1.0f) recorded_samples[i] = 1.0f;
        if (recorded_samples[i] < -1.0f) recorded_samples[i] = -1.0f;
    }

    printf(GET_COLOR(BRIGHT_MAGENTA)"[OYNATMA] İşlenmiş ses çalınıyor...\n"RESET);

    err = Pa_OpenDefaultStream(&stream, 0, NUM_CHANNELS, paFloat32, SAMPLE_RATE, paFramesPerBufferUnspecified, NULL, NULL);
    if (err != paNoError) goto error_play;

    err = Pa_StartStream(stream);
    if (err != paNoError) goto error_play;

    err = Pa_WriteStream(stream, recorded_samples, num_frames);
    if (err != paNoError) goto error_play;

    err = Pa_StopStream(stream);
    if (err != paNoError) goto error_play;
    err = Pa_CloseStream(stream);
    if (err != paNoError) goto error_play;
    stream = NULL;

    printf(GET_COLOR(BRIGHT_MAGENTA)"[OYNATMA] Tamamlandı.\n"RESET);

    SF_INFO sfinfo;
    memset(&sfinfo, 0, sizeof(SF_INFO));
    sfinfo.samplerate = SAMPLE_RATE;
    sfinfo.channels = NUM_CHANNELS;
    sfinfo.format = SF_FORMAT_WAV | SF_FORMAT_FLOAT;

    SNDFILE *outfile = sf_open(RECORDING_FILENAME, SFM_WRITE, &sfinfo);
    if (!outfile) {
        fprintf(stderr, GET_COLOR(RED)"[HATA] '%s' dosyası açılamadı: %s\n"RESET, RECORDING_FILENAME, sf_strerror(NULL));
    } else {
        sf_write_float(outfile, recorded_samples, num_frames);
        sf_close(outfile);
        printf(GET_COLOR(BRIGHT_GREEN)"[KAYIT] İşlenmiş ses '%s' dosyasına kaydedildi.\n"RESET, RECORDING_FILENAME);
    }

    free(recorded_samples);
    return;

error_record:
    if (stream) Pa_CloseStream(stream);
    fprintf(stderr, GET_COLOR(BRIGHT_RED)"PortAudio Kayıt Hatası: %s\n"RESET, Pa_GetErrorText(err));
    free(recorded_samples);
    return;
error_play:
    if (stream) Pa_CloseStream(stream);
    fprintf(stderr, GET_COLOR(BRIGHT_RED)"PortAudio Çalma Hatası: %s\n"RESET, Pa_GetErrorText(err));
    free(recorded_samples);
    return;
}

// ==================
// Ana Menü Fonksiyonu
// ==================
void display_menu() {
    printf("\n%s%s%s\n", GET_COLOR(BRIGHT_YELLOW), BOLD, "========================================"RESET);
    printf("%s%s  SES İŞLEME UYGULAMASI %s\n", GET_COLOR(BRIGHT_CYAN), BOLD, RESET);
    printf("%s%s%s\n", GET_COLOR(BRIGHT_YELLOW), BOLD, "========================================"RESET);
    printf("%s %sSeçenekler:%s\n", GET_COLOR(BRIGHT_WHITE), BOLD, RESET);
    // DEĞİŞİKLİK: "(5 sn kayıt)" ifadesi kaldırıldı
    printf(" %s%s1.%s %sKaydet, İşle, Dinle ve Karışık Hali Kaydet\n",
           GET_COLOR(GREEN), BOLD, RESET, GET_COLOR(BRIGHT_WHITE));
    printf(" %s%s2.%s %sGerçek Zamanlı Ses Değiştirme %s(Mikrofon → Kulaklık)%s\n",
           GET_COLOR(BLUE), BOLD, RESET, GET_COLOR(BRIGHT_WHITE), GET_COLOR(BRIGHT_BLACK), RESET);
    printf(" %s%s3.%s %sÇıkış%s\n",
           GET_COLOR(RED), BOLD, RESET, GET_COLOR(BRIGHT_WHITE), RESET);
    printf("%s%s%s\n", GET_COLOR(BRIGHT_YELLOW), BOLD, "========================================"RESET);
}


// --- DİĞER FONKSİYONLAR (Değişiklik yapılmadı) ---

// ==================
// Realtime Buffer Fonksiyonları
// ==================
void initialize_realtime_buffer(RealtimeBuffer *rb, long size) {
    rb->buffer = (float*) calloc(size, sizeof(float));
    if (!rb->buffer) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"Bellek ayırma hatası!\n"RESET);
        exit(EXIT_FAILURE);
    }
    rb->buffer_size = size;
    rb->read_idx = 0;
    rb->write_idx = 0;
    rb->terminate = false;
    pthread_mutex_init(&rb->mutex, NULL);
    pthread_cond_init(&rb->cond_data_available, NULL);
    pthread_cond_init(&rb->cond_buffer_empty, NULL);
}

void destroy_realtime_buffer(RealtimeBuffer *rb) {
    free(rb->buffer);
    pthread_mutex_destroy(&rb->mutex);
    pthread_cond_destroy(&rb->cond_data_available);
    pthread_cond_destroy(&rb->cond_buffer_empty);
}

bool write_to_buffer(RealtimeBuffer *rb, const float *data, long frames) {
    pthread_mutex_lock(&rb->mutex);
    long current_data_size = (rb->write_idx - rb->read_idx + rb->buffer_size) % rb->buffer_size;
    long free_space = rb->buffer_size - current_data_size - 1;

    while (free_space < frames && !rb->terminate) {
        pthread_cond_wait(&rb->cond_buffer_empty, &rb->mutex);
        current_data_size = (rb->write_idx - rb->read_idx + rb->buffer_size) % rb->buffer_size;
        free_space = rb->buffer_size - current_data_size - 1;
    }

    if (rb->terminate) {
        pthread_mutex_unlock(&rb->mutex);
        return false;
    }

    for (long i = 0; i < frames; ++i) {
        rb->buffer[rb->write_idx] = data[i];
        rb->write_idx = (rb->write_idx + 1) % rb->buffer_size;
    }

    pthread_cond_signal(&rb->cond_data_available);
    pthread_mutex_unlock(&rb->mutex);
    return true;
}

bool read_from_buffer(RealtimeBuffer *rb, float *data, long frames) {
    pthread_mutex_lock(&rb->mutex);
    long current_data_size = (rb->write_idx - rb->read_idx + rb->buffer_size) % rb->buffer_size;

    while (current_data_size < frames && !rb->terminate) {
        pthread_cond_wait(&rb->cond_data_available, &rb->mutex);
        current_data_size = (rb->write_idx - rb->read_idx + rb->buffer_size) % rb->buffer_size;
    }

    if (rb->terminate) {
        pthread_mutex_unlock(&rb->mutex);
        return false;
    }

    for (long i = 0; i < frames; ++i) {
        data[i] = rb->buffer[rb->read_idx];
        rb->read_idx = (rb->read_idx + 1) % rb->buffer_size;
    }

    pthread_cond_signal(&rb->cond_buffer_empty);
    pthread_mutex_unlock(&rb->mutex);
    return true;
}

// ==================
// PortAudio Callback Fonksiyonu
// ==================
int paCallback(const void *inputBufferPtr, void *outputBufferPtr,
               unsigned long framesPerBuffer,
               const PaStreamCallbackTimeInfo* timeInfo,
               PaStreamCallbackFlags statusFlags,
               void *userData) {
    float *out = (float*)outputBufferPtr;
    const float *in = (const float*)inputBufferPtr;

    if (statusFlags & paInputOverflow) fprintf(stderr, GET_COLOR(YELLOW)"[Uyarı-Giriş] Giriş taşması.\n"RESET);
    if (statusFlags & paOutputUnderflow) fprintf(stderr, GET_COLOR(YELLOW)"[Uyarı-Çıkış] Çıkış yetersizliği.\n"RESET);

    if (inputBufferPtr != NULL) {
        if (!write_to_buffer(&inputBuffer, in, framesPerBuffer * NUM_CHANNELS)) {
            return paComplete;
        }
    }

    if (outputBufferPtr != NULL) {
        if (!read_from_buffer(&outputBuffer, out, framesPerBuffer * NUM_CHANNELS)) {
            return paComplete;
        }
    } else {
        for (unsigned int i = 0; i < framesPerBuffer * NUM_CHANNELS; i++) {
            out[i] = 0;
        }
    }

    if (inputBuffer.terminate || outputBuffer.terminate) {
         return paComplete;
    }

    return paContinue;
}

// ==================
// Realtime Processor Thread Fonksiyonu
// ==================
void *realtime_processor_thread(void *arg) {
    (void)arg;
    printf(GET_COLOR(BRIGHT_BLUE)"[GERÇEK ZAMANLI] Ses İşleyici Başlatıldı.\n"RESET);

    float *input_block = (float*) malloc(FRAMES_PER_BUFFER * NUM_CHANNELS * sizeof(float));
    float *processed_block = (float*) malloc(FRAMES_PER_BUFFER * NUM_CHANNELS * sizeof(float));
    if (!input_block || !processed_block) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"Bellek ayırma hatası!\n"RESET);
        exit(EXIT_FAILURE);
    }

    while (!inputBuffer.terminate) {
        if (!read_from_buffer(&inputBuffer, input_block, FRAMES_PER_BUFFER * NUM_CHANNELS)) {
            break;
        }

        simple_pitch_shift(input_block, FRAMES_PER_BUFFER * NUM_CHANNELS, SAMPLE_RATE, PITCH_SHIFT_STEPS);

        for (unsigned int i = 0; i < FRAMES_PER_BUFFER * NUM_CHANNELS; ++i) {
            float noise = ((float)rand() / (float)RAND_MAX) * 0.006f - 0.003f;
            processed_block[i] = input_block[i] + noise;
            if (processed_block[i] > 1.0f) processed_block[i] = 1.0f;
            if (processed_block[i] < -1.0f) processed_block[i] = -1.0f;
        }

        if (!write_to_buffer(&outputBuffer, processed_block, FRAMES_PER_BUFFER * NUM_CHANNELS)) {
            break;
        }
    }

    free(input_block);
    free(processed_block);
    printf(GET_COLOR(BRIGHT_BLUE)"[GERÇEK ZAMANLI] Ses İşleyici Sonlandırıldı.\n"RESET);
    return NULL;
}

// ==================
// Basit Pitch Shift Fonksiyonu
// ==================
void simple_pitch_shift(float *data, long num_frames, int sample_rate, int n_steps) {
    float pitch_factor = powf(2.0f, (float)n_steps / 12.0f);
    float current_sample_pos = 0.0f;
    float *temp_buffer = (float*) calloc(num_frames, sizeof(float));
    if (!temp_buffer) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"Bellek ayırma hatası! Pitch shift yapılamadı.\n"RESET);
        return;
    }

    for (long i = 0; i < num_frames; ++i) {
        long idx1 = (long)floorf(current_sample_pos);
        long idx2 = idx1 + 1;
        float fraction = current_sample_pos - idx1;

        if (idx2 >= num_frames) {
            temp_buffer[i] = data[idx1];
        } else {
            temp_buffer[i] = data[idx1] * (1.0f - fraction) + data[idx2] * fraction;
        }

        current_sample_pos += pitch_factor;

        if (current_sample_pos >= num_frames - 1) {
            for (long j = i + 1; j < num_frames; ++j) {
                temp_buffer[j] = 0.0f;
            }
            break;
        }
    }
    memcpy(data, temp_buffer, num_frames * sizeof(float));
    free(temp_buffer);
}

// ==================
// Mod 2: Gerçek Zamanlı
// ==================
void realtime_mode() {
    PaStream *stream = NULL;
    PaError err;
    pthread_t processor_tid;
    long buffer_frames = SAMPLE_RATE * 2 * NUM_CHANNELS;

    initialize_realtime_buffer(&inputBuffer, buffer_frames);
    initialize_realtime_buffer(&outputBuffer, buffer_frames);

    printf(GET_COLOR(BRIGHT_GREEN)BOLD"*** GERÇEK ZAMANLI SES DEĞİŞTİRME MODU ***"RESET"\n");
    printf(GET_COLOR(BRIGHT_WHITE)"Mikrofon → Anonim Ses (Çıkmak için "GET_COLOR(RED)"Ctrl+C"GET_COLOR(BRIGHT_WHITE)")\n"RESET);

    if (pthread_create(&processor_tid, NULL, realtime_processor_thread, NULL) != 0) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"Thread oluşturma hatası!\n"RESET);
        goto cleanup_realtime;
    }

    err = Pa_OpenDefaultStream(&stream, NUM_CHANNELS, NUM_CHANNELS, paFloat32, SAMPLE_RATE, FRAMES_PER_BUFFER, paCallback, NULL);
    if (err != paNoError) goto error_realtime;

    err = Pa_StartStream(stream);
    if (err != paNoError) goto error_realtime;

    printf(GET_COLOR(BRIGHT_GREEN)"Gerçek zamanlı akış başladı... Konuşun ve işlenmiş sesi duyun.\n"RESET);

    while (Pa_IsStreamActive(stream) == 1) {
        sleep(1);
    }

    printf(GET_COLOR(BRIGHT_RED)"Akış durduruldu.\n"RESET);

    err = Pa_StopStream(stream);
    if (err != paNoError) goto error_realtime;
    err = Pa_CloseStream(stream);
    if (err != paNoError) goto error_realtime;
    stream = NULL;

error_realtime:
    if (err != paNoError && err != paStreamIsStopped) {
        fprintf(stderr, GET_COLOR(BRIGHT_RED)"PortAudio Gerçek Zamanlı Hata: %s\n"RESET, Pa_GetErrorText(err));
    }
    if (stream) {
        Pa_AbortStream(stream);
        Pa_CloseStream(stream);
    }

    inputBuffer.terminate = true;
    outputBuffer.terminate = true;
    pthread_cond_broadcast(&inputBuffer.cond_data_available);
    pthread_cond_broadcast(&inputBuffer.cond_buffer_empty);
    pthread_cond_broadcast(&outputBuffer.cond_data_available);
    pthread_cond_broadcast(&outputBuffer.cond_buffer_empty);

    pthread_join(processor_tid, NULL);

cleanup_realtime:
    destroy_realtime_buffer(&inputBuffer);
    destroy_realtime_buffer(&outputBuffer);
}


// ==================
// Yardımcı Fonksiyon: clear_input_buffer
// ==================
void clear_input_buffer() {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}

File: tr.py
import sounddevice as sd
import numpy as np
import librosa
import queue
import threading
import time
import os
import sys 
try:
    import colorama
    colorama.init() 
except ImportError:
    pass

# ==================
# Renk Kodları
# ==================
class Colors:
    RESET = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

    BLACK = '\033[30m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    WHITE = '\033[37m'

    BRIGHT_BLACK = '\033[90m'
    BRIGHT_RED = '\033[91m'
    BRIGHT_GREEN = '\033[92m'
    BRIGHT_YELLOW = '\033[93m'
    BRIGHT_BLUE = '\033[94m'
    BRIGHT_MAGENTA = '\033[95m'
    BRIGHT_CYAN = '\033[96m'
    BRIGHT_WHITE = '\033[97m'

    BG_BLACK = '\033[40m'
    BG_RED = '\033[41m'
    BG_GREEN = '\033[42m'
    BG_YELLOW = '\033[43m'
    BG_BLUE = '\033[44m'
    BG_MAGENTA = '\033[45m'
    BG_CYAN = '\033[46m'
    BG_WHITE = '\033[47m'

    @staticmethod
    def get_color(color_code):
        """Terminalin renkleri destekleyip desteklemediğini kontrol eder."""
        if sys.stdout.isatty():
            return color_code
        return ""

# ==================
# Genel Ayarlar
# ==================
SAMPLE_RATE = 22050
BLOCK_SIZE = 1024
CHANNELS = 1
DTYPE = 'float32'
RECORDING_FILENAME = "kaydedilen_karisik_ses.wav"

# ==================
# 1. KAYDET → İŞLE → DİNLE → KAYDET Modu
# ==================

def record_process_play_save_mode():
    s = input("Kaç saniye kayıt yapılsın : ")
    DURATION = s  # saniye

    print(f"{Colors.get_color(Colors.BRIGHT_CYAN)}[KAYIT] Konuşmaya başlayın ({{}} saniye)...{Colors.RESET}".format(DURATION))
    recording = sd.rec(int(DURATION * SAMPLE_RATE), samplerate=SAMPLE_RATE, channels=CHANNELS, dtype=DTYPE)
    sd.wait()
    print(f"{Colors.get_color(Colors.BRIGHT_CYAN)}[KAYIT] Tamamlandı.{Colors.RESET}")

    print(f"{Colors.get_color(Colors.BRIGHT_YELLOW)}[İŞLEME] Ses işleniyor...{Colors.RESET}")
    processed_1d = librosa.effects.pitch_shift(recording.flatten(), sr=SAMPLE_RATE, n_steps=-4)
    noise = np.random.normal(0, 0.005, processed_1d.shape).astype(DTYPE)
    processed_1d = processed_1d + noise
    processed_1d = np.clip(processed_1d, -1.0, 1.0)

    print(f"{Colors.get_color(Colors.BRIGHT_MAGENTA)}[OYNATMA] İşlenmiş ses çalınıyor...{Colors.RESET}")
    sd.play(processed_1d, SAMPLE_RATE)
    sd.wait()
    print(f"{Colors.get_color(Colors.BRIGHT_MAGENTA)}[OYNATMA] Tamamlandı.{Colors.RESET}")

    try:
        import soundfile as sf
        sf.write(RECORDING_FILENAME, processed_1d, SAMPLE_RATE)
        print(f"{Colors.get_color(Colors.BRIGHT_GREEN)}[KAYIT] İşlenmiş ses '{RECORDING_FILENAME}' dosyasına kaydedildi.{Colors.RESET}")
    except ImportError:
        print(f"{Colors.get_color(Colors.RED)}[HATA] soundfile kütüphanesi yüklü değil. 'pip install soundfile' ile yükleyiniz. Kayıt yapılamadı.{Colors.RESET}")
    except Exception as e:
        print(f"{Colors.get_color(Colors.RED)}[HATA] Ses kaydedilirken bir sorun oluştu: {e}{Colors.RESET}")

# ==================
# 2. GERÇEK ZAMANLI Mod
# ==================

input_q = queue.Queue()
output_q = queue.Queue()

def audio_input_callback(indata, frames, time_info, status):
    if status:
        print(f"{Colors.get_color(Colors.YELLOW)}[Uyarı-Giriş]{Colors.RESET} {status}")
    input_q.put(indata.copy())

def audio_output_callback(outdata, frames, time_info, status):
    if status:
        print(f"{Colors.get_color(Colors.YELLOW)}[Uyarı-Çıkış]{Colors.RESET} {status}")
    try:
        data = output_q.get_nowait()
        if len(data) >= len(outdata):
            outdata[:] = data[:len(outdata)]
            if len(data) > len(outdata):
                output_q.put_nowait(data[len(outdata):])
        else:
            outdata.fill(0)
    except queue.Empty:
        outdata.fill(0)

def realtime_processor():
    print(f"{Colors.get_color(Colors.BRIGHT_BLUE)}[GERÇEK ZAMANLI] Ses İşleyici Başlatıldı.{Colors.RESET}")
    while True:
        input_data = input_q.get()
        processed_1d = librosa.effects.pitch_shift(input_data.flatten(), sr=SAMPLE_RATE, n_steps=-4)
        noise = np.random.normal(0, 0.003, processed_1d.shape).astype(DTYPE)
        processed_1d = processed_1d + noise
        processed_1d = np.clip(processed_1d, -1.0, 1.0)
        processed_2d = processed_1d.reshape(-1, CHANNELS)
        output_q.put(processed_2d)

def realtime_mode():
    print(f"{Colors.get_color(Colors.BRIGHT_GREEN)}{Colors.BOLD}*** GERÇEK ZAMANLI SES DEĞİŞTİRME MODU ***{Colors.RESET}")
    print(f"{Colors.get_color(Colors.BRIGHT_WHITE)}Mikrofon → Anonim Ses (Çıkmak için {Colors.RED}Ctrl+C{Colors.BRIGHT_WHITE})")

    processor_thread = threading.Thread(target=realtime_processor, daemon=True)
    processor_thread.start()

    with sd.InputStream(callback=audio_input_callback,
                        samplerate=SAMPLE_RATE,
                        channels=CHANNELS,
                        dtype=DTYPE,
                        blocksize=BLOCK_SIZE) as input_stream, \
         sd.OutputStream(callback=audio_output_callback,
                         samplerate=SAMPLE_RATE,
                         channels=CHANNELS,
                         dtype=DTYPE,
                         blocksize=BLOCK_SIZE) as output_stream:
        print(f"{Colors.get_color(Colors.BRIGHT_GREEN)}Gerçek zamanlı akış başladı... Konuşun ve işlenmiş sesi duyun.{Colors.RESET}")
        while input_stream.is_active and output_stream.is_active:
            time.sleep(0.1)
        print(f"{Colors.get_color(Colors.RED)}Akış durduruldu.{Colors.RESET}")

# ==================
# Ana Menü
# ==================

def display_menu():
    c = Colors # Kodu daha kısa yazmak için kısaltma
    print(f"\n{c.get_color(c.BRIGHT_YELLOW)}{c.BOLD}{'='*40}{c.RESET}")
    print(f"{c.get_color(c.BRIGHT_CYAN)}{c.BOLD}{' ':10}SES İŞLEME UYGULAMASI{' ':10}{c.RESET}")
    print(f"{c.get_color(c.BRIGHT_YELLOW)}{c.BOLD}{'='*40}{c.RESET}")

    print(f"{c.get_color(c.BRIGHT_WHITE)} {c.BOLD}Seçenekler:{c.RESET}")
    print(f" {c.get_color(c.GREEN)}{c.BOLD}1.{c.RESET} {c.BRIGHT_WHITE}Kaydet, İşle, Dinle ve Karışık Hali Kaydet {c.RESET}{c.get_color(c.BRIGHT_BLACK)}(5 sn kayıt){c.RESET}")
    print(f" {c.get_color(c.BLUE)}{c.BOLD}2.{c.RESET} {c.BRIGHT_WHITE}Gerçek Zamanlı Ses Değiştirme {c.RESET}{c.get_color(c.BRIGHT_BLACK)}(Mikrofon → Kulaklık){c.RESET}")
    print(f" {c.get_color(c.RED)}{c.BOLD}3.{c.RESET} {c.BRIGHT_WHITE}Çıkış{c.RESET}")
    print(f"{c.get_color(c.BRIGHT_YELLOW)}{c.BOLD}{'='*40}{c.RESET}")

def main():
    while True:
        display_menu()
        choice = input(f"{Colors.get_color(Colors.BRIGHT_WHITE)}Lütfen bir seçenek girin {Colors.get_color(Colors.CYAN)}(1-3){Colors.BRIGHT_WHITE}: {Colors.RESET}")

        if choice == '1':
            print(f"\n{Colors.get_color(Colors.MAGENTA)}{Colors.BOLD}>>> Seçim: Kaydet & İşle Modu <<<{Colors.RESET}\n")
            record_process_play_save_mode()
            print(f"\n{Colors.get_color(Colors.BRIGHT_YELLOW)}{'='*50}{Colors.RESET}\n")
        elif choice == '2':
            print(f"\n{Colors.get_color(Colors.MAGENTA)}{Colors.BOLD}>>> Seçim: Gerçek Zamanlı Mod <<<{Colors.RESET}\n")
            try:
                realtime_mode()
            except KeyboardInterrupt:
                print(f"\n{Colors.get_color(Colors.RED)}Gerçek zamanlı moddan çıkılıyor.{Colors.RESET}")
            except Exception as e:
                print(f"{Colors.get_color(Colors.RED)}[HATA] Bir hata oluştu: {e}{Colors.RESET}")
            print(f"\n{Colors.get_color(Colors.BRIGHT_YELLOW)}{'='*50}{Colors.RESET}\n")
        elif choice == '3':
            print(f"{Colors.get_color(Colors.BRIGHT_RED)}Programdan çıkılıyor. Hoşça kalın!{Colors.RESET}")
            break
        else:
            print(f"{Colors.get_color(Colors.RED)}Geçersiz seçenek. Lütfen {Colors.CYAN}1, 2{Colors.RED} veya {Colors.RED}3{Colors.RESET}{Colors.RED} girin.{Colors.RESET}")
            time.sleep(1)

if __name__ == "__main__":
    main()